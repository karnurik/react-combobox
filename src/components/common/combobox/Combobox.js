import React, { Component } from 'react';
import ComboInput from './combo-input/ComboInput';
import InputSelections from './input-selections/InputSelections';

class Combobox extends Component {
    constructor(props) {
        super(props);
        this.state = {
          selectedOptions : [],
          searching       : false
        }
    }

    /* 
      * @description - Function toggles seraching state, gets called when user clicks on input box for selection
      * @param event 
      * @return na
    */ 
    handleInputFocus = e => {
        this.setState({searching: !this.state.searching});
    }

    /* 
      * @description - Function adds mousedown event listners and calls handleClickOutside function
      * @param - na 
      * @return na
    */ 
    componentDidMount() {
        document.addEventListener('mousedown', this.handleClickOutside);
    }

    /* 
      * @description - Function sets wrapper ref on the called element
      * @param - node (current event node) 
      * @return na
    */ 
    setWrapperRef = node => { 
        this.wrapperRef = node 
    }

    /* 
      * @description - Function handles outside clicks and sets the searching state false
      * @param - event  
      * @return na
    */ 
    handleClickOutside = e => {
        if (this.wrapperRef && !this.wrapperRef.contains(e.target)) {
            this.setState({searching: false});
        }
    }

    /* 
      * @description - Function creates an array with the selected options
      * @param - selectedId 
      * @return na
    */ 
    handleSelectedOptions = selectedId => {
        const selectedOptions = this.state.selectedOptions ? this.state.selectedOptions : [];
        
        (selectedOptions && selectedOptions.includes(selectedId)) ? 
            selectedOptions.splice(selectedOptions.indexOf(selectedId) , 1)
            :
            selectedOptions.push(selectedId);
        

        this.setState({selectedOptions: selectedOptions});
    }

    /* 
      * @description - Function removes mousedown event listners and calls handleClickOutside function
      * @param - na 
      * @return na
    */ 
    componentWillUnmount() {
        document.removeEventListener('mousedown', this.handleClickOutside);
    }

    render() {
        const selectedOptions = this.state.selectedOptions ? this.state.selectedOptions : [];
        const data            = this.props.data ? this.props.data : [];
        const label           = this.props.label ? this.props.label : '';
        
        /**
         * code to check the exception handling pass value debug to label 
         */
        if(label == 'debug'){
            throw new Error('I crashed!');
        }
        
        return (
            <div className='combo-box' ref={this.setWrapperRef} >
                 <label>{label}</label>
                 {this.state.searching ? 
                    <ComboInput data={data} 
                        selectedOptions={selectedOptions} 
                        handleSelectedOptions={this.handleSelectedOptions} 
                        searchPlaceholder={this.props.searchPlaceholder}
                    /> : 
                    <InputSelections data={data} 
                        selectedOptions={selectedOptions} 
                        handleInputFocus={this.handleInputFocus}
                        placeholder= {this.props.placeholder}
                    /> }
            </div>
        );
    }
}

export default Combobox;