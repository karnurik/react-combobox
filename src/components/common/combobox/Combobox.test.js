import React from 'react';
import Combobox from './combobox';
import renderer from 'react-test-renderer';

const data = [
  { id: '1', name: 'Candy Eat' },
  { id: '2', name: 'Onions Pick' },
  { id: '3', name: 'Fanny Pick' },
  { id: '4', name: 'Click o Wisp' }
];
const placeholder = 'Select Games';
const searchPlaceholder = 'Search Games';
const label = 'Games';

test('Combobox component renders correctly', () => {
  const component = renderer.create(
    <Combobox
      data={data}
      placeholder={placeholder}
      searchPlaceholder={searchPlaceholder}
      label={label}
    />
  );

  expect(component.toJSON()).toMatchSnapshot();
});
