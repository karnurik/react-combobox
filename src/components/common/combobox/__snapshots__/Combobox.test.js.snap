// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combobox component renders correctly 1`] = `
<div
  id="outer"
>
   
  <div
    className="combo-box"
  >
    <label>
      Games
    </label>
    <div
      className="drop-down"
      onClick={[Function]}
    >
      <input
        className="form-input"
        defaultValue=""
        placeholder="Select Games"
        type="text"
      />
      <i
        className="down-arrow"
      />
    </div>
  </div>
</div>
`;
